Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -69,8 +69,60 @@
 
 #if HAVE(WEB_TRANSPORT)
 
-static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&& connectionToWebProcess, URL&& url, WebKit::WebPageProxyIdentifier&& pageID, WebCore::ClientOrigin&& clientOrigin, sec_trust_t trust, sec_protocol_verify_complete_t completion)
+static bool leafCertificateMatchesWebTransportHash(sec_trust_t trust, const Vector<WebCore::WebTransportHash>& hashes)
 {
+    // https://www.w3.org/TR/webtransport/#verify-a-certificate-hash
+    RetainPtr secTrust = adoptCF(sec_trust_copy_ref(trust));
+    if (!secTrust)
+        return false;
+
+    RetainPtr chain = adoptCF(SecTrustCopyCertificateChain(secTrust.get()));
+    if (!chain || !CFArrayGetCount(chain.get()))
+        return false;
+
+    RetainPtr leafCertificate = checked_cf_cast<SecCertificateRef>(CFArrayGetValueAtIndex(chain.get(), 0));
+    if (!leafCertificate)
+        return false;
+
+    RetainPtr x509Data = adoptNS(bridge_cast(SecCertificateCopyData(leafCertificate.get())));
+    if (!x509Data)
+        return false;
+
+    std::array<uint8_t, CC_SHA256_DIGEST_LENGTH> sha2;
+    CC_SHA256([x509Data bytes], [x509Data length], sha2.data());
+
+    bool hashMatches = std::ranges::any_of(hashes, [&] (auto& hash) {
+        return hash.algorithm == "sha-256"_s && equalSpans(std::span(sha2), hash.value.span());
```

Comment:
Does/should the `hash.algorithm` check be case-insensitive? or it must match exactly?